### PR TITLE
Extend options system to handle a set of options for each feature

### PR DESCRIPTION
**Pull request by RobPavey**
This PR changes the design of how features are registered with the option system and it switches the AGC feature to use the new system and adds all the AGC options to the UI.

If this is approved the rest of the features can be changed to use the new system to check whether they are enabled in the options.

There are further improvements that could be made but I don't want to make this PR any more complicated.

Changes to note:
* The features array has moved from being created and filled in options.js to being created in src/core/options/options_registry.mjs and filled by calling registerFeature (either in src/features/register_feature_options.mjs directly or in a file included from there)
* When a feature is registered the given object can have an options property that defines all the options
* The code in options.js that builds the options page now recursively walks the options for the feature and builds the UI
* The system handles default values for options. This could be extended to handle default values for the feature enable flags also. It also handle default values for the options in tests (can be useful if the default is changed but you don't want to invalidate regression data)
* The code used in the content code to read the options from storage is now in src/core/options/options_storage.mjs
Only AGC has been switched over to this so far but it means that the feature code no longer has to use chrome.storage.sync.get
* When a feature is enabled or disabled it only makes one call to chrome.storage.sync.set.
The existing code was getting errors because it was making too many calls to this if you flip a few sliders within 1 minute.

### CODE REVIEW

**Comment by RobPavey** [comment]
The tests are failing because of conflict between the way the new jest code is working and the way my changes work.
There are actually 2 issues:
1. There was a merge error in the agc test file that got renamed in the jest changes. This is easy to fix.
2. It doesn't allow an import statement in agc.test.js. Strangely other imports statements are working - perhaps it doesn't like an import of the form `import "../../features/register_feature_options.mjs";` in a .js file but allows it in a .mjs file. This will take some investigation.

**Comment by RobPavey** [comment]
I got the tests working but had to modify the jest config to allow ES modules.

**Comment by RobPavey** [comment]
One design decision we can discuss (though it doesn't need to hold up this PR) is about how the content script code loads the options from storage and checks them in the features. The original design (and this new design) do a chrome.storage.sync.get at the time that the feature check is done. If think the change @PeWu made changed this so that the options were read once when the content loaded.

There are pros and cons to each. The latter means less calls to chrome.storage.sync.get (may not be significant as it is cached) but it means that, if a feature is enabled or disabled, a page reload is always required for that to take effect. In most cases (because the feature is adding something to the page on page load) the reload is required anyway. But there could be some features, say something that runs when you click the preview button, that can make the check at that time so that they don't need a reload when the feature is enabled or disabled.

It is a pretty minor issue that probably wouldn't be a big deal to users.

For feature options on the other hand it is important that a page reload is not required when options are changed. For AGC for example you can press the AGC button, then press it again to undo, change the options, then press the AGC button again and the new options will be used.

**Review by shogenapps — APPROVED** [review]
I've got this working for me. 
I'm sure Rob's instructions are great, so you can think of this as a note to self, but here's what I did for dark mode (working on Rob's fork):

1. Add `import { checkIfFeatureEnabled, getFeatureOptions } from "../../core/options/options_storage.mjs"` to darkMode.js.
2. Change the chrome.sync call to `checkIfFeatureEnabled("darkMode").then((result) => {
  if (result) {
    initDarkMode();  
  }
});`
3. For options, create darkMode_options.mjs. 
This file looks like this:

```
import {
  registerFeature,
  OptionType,
} from "../../core/options/options_registry.mjs";

const darkModeFeature = {
  name: "Dark Mode",
  id: "darkMode",
  description: "Make WikiTree dark.",
  category: "Style",
  options: [
    {
      id: "mode",
      type: OptionType.RADIO,
      label: "When",
      values: [
        {
          value: "always",
          text: "Always",
        },
        {
          value: "system",
          text: "Follow my system preference",
        },
      ],
      defaultValue: "always",
    },
  ],
};

registerFeature(darkModeFeature);
```


4. Add `import "./darkMode/darkMode_options.mjs";` to features/register_feature_options.mjs

I hope I didn't miss anything.

File: src/options.js [hunk 1]
```diff
@@ -1,133 +1,8 @@
...
-      "When attaching a person by ID, you can see some details of the person and check that you've entered the correct ID.",
-    category: "Editing",
-  },
-];
+import { features, OptionType } from "./core/options/options_registry.mjs";
+import "./features/register_feature_options.mjs";
+
 
 // Categories
 const categories = ["Global", "Profile", "Editing", "Style"];
```
Comment (PeWu) [inline src/options.js:8]: The "Other" category must have gotten removed accidentally

Comment (RobPavey) [reply src/options.js:8]: I don't think it was ever there. If you look at the lines below that:
```
// If a new feature is added with a new category, add the category to the list
features.forEach(function (feature) {
  if (!categories.includes(feature.category)) {
    categories.push(feature.category);
  }
});
```
It has always handled categories that are not in the initial initialization of `categories`

Comment (PeWu) [reply src/options.js:8]: This commit added "Other":
https://github.com/wikitree/wikitree-browser-extension/commit/93dc7faf41488520cf4bd1fb5185ae217078d5bc
@shogenapps

Comment (RobPavey) [reply src/options.js:8]: Oh, yes. It was between your changes and mine. It works fine as it is for now. We can decide if we want to move to a different system of treating the category like an enum. We may also want to separate the category id from the category text/description. (see my notes on https://github.com/wikitree/wikitree-browser-extension/issues/87)

**Pull request merged**